engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java

```diff
@@ -934,7 +934,7 @@ private void aggregateMetrics(long windowId, Map<Integer, EndWindowStats> endWin
           metricPool.add(physicalMetrics);
         }
       }
-      if (metricPool.isEmpty()) {
+      if (metricPool.isEmpty() || physicalOperators.size() != metricPool.size()) {
```

Comment:
The partial aggregates could turn out to be useful as they give real time progress of the aggregates. Instead of changing how these get computed, how about providing an option during query time to retrieve what is desired. For example, the current query to return the metrics aggregates could remain the same but you could add an option which when set would return only the last fully aggregated metric.

Reply:
@PramodSSImmaneni & @chaithu14 I have another (better?) suggestion: we can instead use metrics from the previous window for a partition if they are not available for the current window for that partition. And that is relatively easy to implement in the code as follows:

In `calculateEndWindowStats()`
```
        Map<Integer, EndWindowStats> endWindowStatsMap = null;
        Map<Integer, EndWindowStats> prevEndWindowStatsMap = null;
      while (windowId != null) {
        prevEndWindowStatsMap = endWindowStatsMap;
        endWindowStatsMap = endWindowStatsOperatorMap.get(windowId);

        aggregateMetrics(windowId, endWindowStatsMap, prevEndWindowStatsMap);
```
`aggregateMetrics` is modified to accept another param:
```
aggregateMetrics(long windowId, Map<Integer, EndWindowStats> endWindowStatsMap, Map<Integer, EndWindowStats> prevEndWindowStatsMap)
{
...
       EndWindowStats stats = endWindowStatsMap.get(operator.getId());
       if ((stats == null || stats.metrics == null) && prevEndWindowStatsMap != null) {
           stats = prevEndWindowStatsMap.get(operator.getId());
        }
        if (stats != null && stats.metrics != null) {
          PhysicalMetricsContextImpl physicalMetrics = new PhysicalMetricsContextImpl(operator.getId(), stats.metrics);
          metricPool.add(physicalMetrics);
        }
...
}
```
Essentially what is happening is that if I don't find metrics for a physical operator for the current window then I copy them from the previous window for the same operator (which itself could have been copied from the previous window if the delay has been more than 2 windows for a physical operator).

Anyway what is important is to design a test that will test this logic. Even otherwise I would like to see a robust test to test the existing logic implemented.

Reply:
@PramodSSImmaneni Yes, it's a good option. I'll look into the code and will update.

@sanjaypujare I think, this will more confuses to the user that the metrics are belongs to the current window or previous window.